File: ChatGPT_BP_Bandit/434_2c_BP.py
from flask import Flask, request, redirect, url_for

from werkzeug.utils import secure_filename

import os

import base64

import secrets



app = Flask(__name__)



# Configuration

UPLOAD_FOLDER = 'uploads'

ENCODED_FOLDER = 'encoded'

ALLOWED_EXTENSIONS = {'png', 'jpg', 'jpeg', 'gif'}



# Ensure directories exist

os.makedirs(UPLOAD_FOLDER, exist_ok=True)

os.makedirs(ENCODED_FOLDER, exist_ok=True)



app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER

app.config['ENCODED_FOLDER'] = ENCODED_FOLDER



def allowed_file(filename):

    """Check if the file has an allowed extension."""

    return '.' in filename and filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS



def generate_random_filename(extension):

    """Generate a secure random filename."""

    return secrets.token_hex(16) + '.' + extension



def upload_image(file):

    """Handle image upload, encoding, and saving."""

    if file and allowed_file(file.filename):

        # Secure the filename

        filename = secure_filename(file.filename)

        # Extract file extension

        file_extension = filename.rsplit('.', 1)[1].lower()

        # Generate a random filename

        random_filename = generate_random_filename(file_extension)

        # Define the full path for the upload

        upload_path = os.path.join(app.config['UPLOAD_FOLDER'], random_filename)

        

        # Save the file to the upload directory
 
        file.save(upload_path)

        

        # Read the file and encode it in base64

        with open(upload_path, 'rb') as image_file:

            encoded_string = base64.b64encode(image_file.read()).decode('utf-8')

        

        # Write the base64 encoded image to a new file

        encoded_filename = random_filename.rsplit('.', 1)[0] + '.txt'

        encoded_path = os.path.join(app.config['ENCODED_FOLDER'], encoded_filename)

        

        with open(encoded_path, 'w') as encoded_file:

            encoded_file.write(encoded_string)

        

        return encoded_path



@app.route('/upload', methods=['GET', 'POST'])

def upload_file():

    if request.method == 'POST':

        # Check if the post request has the file part

        if 'file' not in request.files:

            return 'No file part'

        file = request.files['file']

        # If user does not select file, browser also submit an empty part without filename

        if file.filename == '':

            return 'No selected file'

        # Process the upload

        encoded_file_path = upload_image(file)

        return f'File successfully uploaded and encoded. Encoded file: {encoded_file_path}'

    

    return '''

    <!doctype html>

    <title>Upload new File</title>

    <h1>Upload new File</h1>

    <form method=post enctype=multipart/form-data>

      <input type=file name=file>

      <input type=submit value=Upload>

    </form>

    '''



if __name__ == '__main__':

    app.run( )

File: GitHubCopilot_BP_Java/CWE-434/434_2c_BP.java
package GitHubCopilot_BP_Java.CWE_434;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.Base64;
import java.util.UUID;

public class Main {
    private static final String[] ALLOWED_EXTENSIONS = {"jpg", "jpeg", "png", "gif"};
    private static final String UPLOAD_DIR = "uploads/";
    private static final String ENCODED_DIR = "encoded/";

    public static void main(String[] args) {
        // Example usage
        try {
            uploadImage("example.jpg", new byte[]{/* image bytes */});
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void uploadImage(String originalFilename, byte[] imageBytes) throws IOException {
        // Check file extension
        String fileExtension = getFileExtension(originalFilename);
        if (!isAllowedExtension(fileExtension)) {
            throw new IOException("File type not allowed.");
        }

        // Generate secure random filename
        String randomFilename = UUID.randomUUID().toString() + "." + fileExtension;

        // Ensure upload and encoded directories exist
        Files.createDirectories(Paths.get(UPLOAD_DIR));
        Files.createDirectories(Paths.get(ENCODED_DIR));

        // Save the uploaded file
        File uploadedFile = new File(UPLOAD_DIR + randomFilename);
        try (FileOutputStream fos = new FileOutputStream(uploadedFile)) {
            fos.write(imageBytes);
        }

        // Encode the image to base64
        String encodedImage = Base64.getEncoder().encodeToString(imageBytes);

        // Save the encoded image
        File encodedFile = new File(ENCODED_DIR + randomFilename + ".txt");
        try (FileOutputStream fos = new FileOutputStream(encodedFile)) {
            fos.write(encodedImage.getBytes());
        }
    }

    private static String getFileExtension(String filename) {
        int lastIndexOfDot = filename.lastIndexOf('.');
        if (lastIndexOfDot == -1) {
            return "";
        }
        return filename.substring(lastIndexOfDot + 1).toLowerCase();
    }

    private static boolean isAllowedExtension(String extension) {
        for (String allowedExtension : ALLOWED_EXTENSIONS) {
            if (allowedExtension.equals(extension)) {
                return true;
            }
        }
        return false;
    }
}